nucypher/policy/kits.py

```diff
@@ -0,0 +1,249 @@
+
+class MessageKit:
+    """
+    All the components needed to transmit and verify an encrypted message.
+    """
+
+    @classmethod
+    def author(cls,
+               recipient_key: PublicKey,
+               plaintext: bytes,
+               signer: 'SignatureStamp',
+               sign_plaintext: bool = True
+               ) -> 'MessageKit':
+        if signer is not DO_NOT_SIGN:
+            # The caller didn't expressly tell us not to sign; we'll sign.
+            if sign_plaintext:
+                # Sign first, encrypt second.
+                sig_header = SIGNATURE_TO_FOLLOW
+                signature = signer(plaintext)
+                capsule, ciphertext = umbral.encrypt(recipient_key, sig_header + bytes(signature) + plaintext)
+                signature_in_kit = None
+            else:
+                # Encrypt first, sign second.
+                sig_header = SIGNATURE_IS_ON_CIPHERTEXT
+                capsule, ciphertext = umbral.encrypt(recipient_key, sig_header + plaintext)
+                signature = signer(ciphertext)
+                signature_in_kit = signature
+            message_kit = cls(ciphertext=ciphertext,
+                              capsule=capsule,
+                              sender_verifying_key=signer.as_umbral_pubkey(),
+                              signature=signature_in_kit)
+        else:
+            # Don't sign.
+            sig_header = NOT_SIGNED
+            capsule, ciphertext = umbral.encrypt(recipient_key, sig_header + plaintext)
+            message_kit = cls(ciphertext=ciphertext, capsule=capsule)
+
+        return message_kit
+
+    def __init__(self,
+                 capsule: Capsule,
+                 ciphertext: bytes,
+                 sender_verifying_key: Optional[PublicKey] = None,
+                 signature: Optional[Signature] = None,
+                 ):
+        self.ciphertext = ciphertext
+        self.capsule = capsule
+        self.sender_verifying_key = sender_verifying_key
+        self.signature = signature
+
+    def __eq__(self, other):
+        return (self.ciphertext == other.ciphertext and
+                self.capsule == other.capsule and
+                self.sender_verifying_key == other.sender_verifying_key and
+                self.signature == other.signature)
+
+    def __str__(self):
+        return f"{self.__class__.__name__}({self.capsule})"
+
+    def __bytes__(self):
+        return (bytes(self.capsule) +
+                (b'\x00' if self.signature is None else (b'\x01' + bytes(self.signature))) +
+                (b'\x00' if self.sender_verifying_key is None else (b'\x01' + bytes(self.sender_verifying_key))) +
+                VariableLengthBytestring(self.ciphertext))
+
+    @classmethod
+    def from_bytes(cls, data: bytes):
+        splitter = BytestringSplitter(
+            capsule_splitter,
+            (bytes, 1))
+
+        capsule, signature_flag, remainder = splitter(data, return_remainder=True)
+
+        if signature_flag == b'\x00':
```

Comment:
Nitpick: Could `b'\x00'` and `b'\x01'` be translated into constants, for example: `NO_SIG`, `HAS_SIG`?

Reply:
Fair point, but this code might prove to be unnecessary depending on the decision on #2743; otherwise it is better fixed in BSS (as a generic support for Optional objects). Added TODOs.